docs/source/en/llm_optims.md

```diff
@@ -0,0 +1,324 @@
+from transformers.testing_utils import CaptureLogger
+import torch
+
+prompts = [
+    "Simply put, the theory of relativity states that ",
+    "My favorite all time favorite condiment is ketchup.",
+]
+
+NUM_TOKENS_TO_GENERATE = 40
+torch_device = "cuda"
+
+tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", pad_token="</s>", padding_side="right")
+model = LlamaForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", device_map="sequential")
+inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(model.device)
+
+def decode_one_tokens(model, cur_token, input_pos, cache_position):
+    logits = model(
+        cur_token, position_ids=input_pos, cache_position=cache_position, return_dict=False, use_cache=True
+    )[0]
+    new_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+    return new_token
+```
+
+The are two important things you must do to enable static kv-cache and torch.compile with the `_setup_cache` method:
+
+1. Access the model's `_setup_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.
+
+2. Call torch.compile on the model to compile the forward pass with the static kv-cache.
+
+```py
+batch_size, seq_length = inputs["input_ids"].shape
+with torch.no_grad():
+     model._setup_cache(StaticCache, 2, max_cache_len=4096)
+     cache_position = torch.arange(seq_length, device=torch_device)
+     generated_ids = torch.zeros(
+         batch_size, seq_length + NUM_TOKENS_TO_GENERATE + 1, dtype=torch.int, device=torch_device
+     )
+     generated_ids[:, cache_position] = inputs["input_ids"].to(torch_device).to(torch.int)
+
+     logits = model(**inputs, cache_position=cache_position, return_dict=False, use_cache=True)[0]
+     next_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+     generated_ids[:, seq_length] = next_token[:, 0]
+
+     decode_one_tokens = torch.compile(decode_one_tokens, mode="reduce-overhead", fullgraph=True)
+     cache_position = torch.tensor([seq_length + 1], device=torch_device)
+     for _ in range(1, NUM_TOKENS_TO_GENERATE):
+         with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=False, enable_math=True):
+             next_token = decode_one_tokens(model, next_token.clone(), None, cache_position)
+             generated_ids[:, cache_position] = next_token.int()
+         cache_position += 1
+
+text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
+text
+['Simply put, the theory of relativity states that 1) the speed of light is constant, 2) the speed of light is the same for all observers, and 3) the laws of physics are the same for all observers.',
+ 'My favorite all time favorite condiment is ketchup. I love it on everything. I love it on my eggs, my fries, my chicken, my burgers, my hot dogs, my sandwiches, my salads, my p']
+```
```

Comment:
For both of the compile options - I think it would be really compelling for a user if they had actual numbers e.g. fdo we get a 4x speed up under all cases? Do we get the same speed-up for these two cache approaches?

Reply:
Totally agree, I'll ask @gante about this 🙂 

In the meantime, would it be ok to merge this PR first and add this in a separate PR or would you prefer to wait?

Reply:
Sure!

Reply:
Sure